### ISSUE TITLE
Handle special characters in database URL

### ISSUE THREAD

**Issue by sephii** [kennethreitz/dj-database-url#39]
If the password contains special characters such as ? or /, the following exception is raised (here I tried to use "foo?bar" as a password):

```
Traceback (most recent call last):
  File "./manage.py", line 19, in <module>
    execute_from_command_line(sys.argv)
  File "/var/www/myproject/dev/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 399, in execute_from_command_line
    utility.execute()
  File "/var/www/myproject/dev/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 376, in execute
    sys.stdout.write(self.main_help_text() + '\n')
  File "/var/www/myproject/dev/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 240, in main_help_text
    for name, app in six.iteritems(get_commands()):
  File "/var/www/myproject/dev/env/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 107, in get_commands
    apps = settings.INSTALLED_APPS
  File "/var/www/myproject/dev/env/local/lib/python2.7/site-packages/django/conf/__init__.py", line 54, in __getattr__
    self._setup(name)
  File "/var/www/myproject/dev/env/local/lib/python2.7/site-packages/django/conf/__init__.py", line 49, in _setup
    self._wrapped = Settings(settings_module)
  File "/var/www/myproject/dev/env/local/lib/python2.7/site-packages/django/conf/__init__.py", line 128, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
  File "/var/www/myproject/dev/env/local/lib/python2.7/site-packages/django/utils/importlib.py", line 40, in import_module
    __import__(name)
  File "/var/www/myproject/dev/myproject/myproject/settings/base.py", line 175, in <module>
    'default': dj_database_url.parse(get_env_variable('DATABASE_URL'))
  File "/var/www/myproject/dev/env/local/lib/python2.7/site-packages/dj_database_url.py", line 84, in parse
    'PORT': url.port or '',
  File "/usr/lib/python2.7/urlparse.py", line 110, in port
    port = int(port, 10)
ValueError: invalid literal for int() with base 10: 'foo'
```

I tried to escape the special characters with backslashes but then it got interpreted as "foo\\". I also tried to url-encode them, ie. replacing "?" by %63 but then the password returned by dj-database-url was "foo%63bar".

The issue can be easily reproduced with the following URL: `postgres://foo:foo?bar@localhost/foobar`.

Maybe I'm just missing something but if that's the case I think it would be worth documenting it somewhere.

**Comment by rapilabs** [comment]
Interestingly enough, '@' appearing in the password unescaped works fine, yet '#' will be used as the fragment delimiter (Python 2.7.9).

**Comment by levivm** [comment]
+1

**Comment by kennethreitz** [comment]
Done!

**Comment by sephii** [comment]
Wow thanks, I didn't think this would get fixed someday! Could you please point me to the commit that fixed the bug? I'd be interested to know what the bug was.

**Comment by kennethreitz** [comment]
Ah, I apologize, special charecters are handled by in another part of the URL now. 

The part of the URL in question is called the **scheme authorization part**, and characters here are always to be precent-encoded, in order to be a valid URL.

**Comment by iMerica** [comment]
This issue persists for me in `dj-database-url==0.4.2`

**Comment by iMerica** [comment]
**Steps to Reproduce**
1) Export the following as an env var 
`export DB_URL=postgresql://foo:123XYZ##ABC%@example.com:5432/test_db`
2) Use Django db URL and parse the URL above: `dj_database_url.parse(os.environ.get('DB_URL'))` 

**Expected Results**
dj-database-url should either parse this properly or warn me with a helpful error if theres an issue

**Actual results**
dj-database-url raises the error `ValueError: invalid literal for int() ...` because dj-database-url sees the `#` as a scheme delimiter or something.


I should be able to work around this by encoding before parsing. Just leaving some details here in case it helps 😃 .

**Comment by slonokot** [comment]
the issue is still reproduced in version 0.5.0 see @iMerica's  post above (python 3.7.0):

# export DB_URL=postgresql://foo:123XYZ##ABC%@example.com:5432/test_db
# python3
Python 3.7.0 (default, Jul 23 2018, 20:22:55)
[Clang 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os, dj_database_url
>>> dj_database_url.parse(os.environ.get('DB_URL'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/username/code/test/djdatabase/lib/python3.7/site-packages/dj_database_url.py", line 105, in parse
    port = (str(url.port) if url.port and engine == SCHEMES['oracle']
  File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/parse.py", line 169, in port
    port = int(port, 10)
ValueError: invalid literal for int() with base 10: '123XYZ'

**Comment by kkaczmarczyk** [comment]
Yeah I've got the same.

**Comment by Joonalai** [comment]
Also character `#` produces the error.

**Comment by rahulshw** [comment]
I still see this issue in version 0.5.0 .

**Comment by marko-kostic-scribe** [comment]
I have observed this issue in the version `2.0.0`. Using the `urlquote` workaround does work for me.

A code snippet for the poor souls from the future:

```
#!/usr/bin/env python3
import dj_database_url
import os
from urllib.parse import quote as urlquote

DB_URL = urlquote(os.environ.get("DATABASE_URL", "postgresql://"), ":/@")
DATABASES["default"] = dj_database_url.parse(DB_URL) # Django specific
```

**Comment by stuaxo** [comment]
To help others using mssql -

I hit this trying to connect from an mssql database, and this wasn't the cause, just how a bug manifested (he cause was not the special characters in the password as I expected.(

dj_database_url remaps the engine type - there seem to be many (legacy?) odbc drivers, in the

`SCHEMES` dictionary in `dj_database_url/__init__.py`

My url started with `mssql` and so I expected the ENGINE would be `mssql`, however SCHEMES maps `mssql` to  `sql_server.pyodbc`.

This was different to the ENGINE `mssql` I knew worked from manually setting the database dict.

Once I set my url to start with `mssqlms://` (which maps to ENGINE `mssql`) things started working.

I verified this by setting up my DATABASE setting as a dict, then putting a debugger at the end of dj_database_url.parse to check the actual settings being used.
